### DISCUSSION TITLE
Project optimization suggestions and new feature requests

### DISCUSSION THREAD

**Discussion by anncwb** [vbenjs/vue-vben-admin#251] [Show and tell]
If you have good suggestions or ideas, please feel free to mention them here.

If the system does not have the function you want, you can put it here, and we will develop it according to the situation!

Function proposal format:

```
Name: Add xxx function
Role: Business scenario description
Online reference example (if available): https://xxxx
```

**Comment by anncwb** [comment]
History discussion board association: https://github.com/anncwb/vue-vben-admin/issues/13 https://github.com/anncwb/vue-vben-admin/issues/74

**Comment by wsxe9988** [comment]
![image](https://user-images.githubusercontent.com/5828057/111973850-e23d5880-8b39-11eb-883a-8cca717710ac.png)
能否在红色圈出的地方把修改密码 个人中心 个人设置 移过去呢

**Reply by winthesky** [reply]
你们太懒了

**Comment by wsxe9988** [comment]
![image](https://user-images.githubusercontent.com/5828057/111973967-fda86380-8b39-11eb-970b-90d132777c17.png)
能否增加 开启 Tagsview 图标设置，开启 Tagsview 拖拽设置，开启水印设置 水印图案设置 Tagsview 风格的切换设置，菜单字体背景高亮设置，菜单背景渐变设置，顶栏背景渐变设置

**Comment by wsxe9988** [comment]
![image](https://user-images.githubusercontent.com/5828057/111987648-6e577c00-8b4a-11eb-985b-1a03d256be1a.png)
能否在空的时候写个这种过滤筛选的demo

**Comment by wsxe9988** [comment]
![image](https://user-images.githubusercontent.com/5828057/111987785-99da6680-8b4a-11eb-88f5-d62460348e75.png)
![image](https://user-images.githubusercontent.com/5828057/111987832-a65ebf00-8b4a-11eb-9011-26bf7179a5e4.png)
 空的时候希望可以 弄个表单自适应的demo

**Comment by i7eo** [comment]
类似 antd-prop 的 ProDescription 组件，是否有开发计划？

**Comment by Miofly** [comment]
Name: 多种筛选功能集合的一种组件（企查查等都有这种平铺的筛选组件)
Role: ![image](https://user-images.githubusercontent.com/42859910/147723266-61f84594-7d35-4677-b8ab-688534811a25.png)
Online reference example (if available): https://www.qyyjt.cn/cloudTotal

**Comment by Miofly** [comment]
Name: 每一个 tab 页面上方进度条是否展示的控制
Role: 希望对 axios 加入一层封装，
1.如对某个接口加入 参数 showProess: true，则这个接口没请求完之前页面上方保持有进度条，如果一个 tab 页面有多个接口同时调用，所有接口都请求完成上方进度条再消失
2.对进度条参数加入是否每次调用都显示进度条，如showProess: { once: true }, once  为 true 代表只有首次调用接口才加载进度条， false 代表此接口只要调用，tab 页面都会出现进度条

**Comment by Miofly** [comment]
Name: 多个 vben 后台管理项目，如何把共同的东西抽离出来
Role:
1.假设有 10-20 个 vben 这种后台管理，如何对共通的地方进行抽离，比如 uitls、components、hooks等
还有就是一些配置文件，包括 eslint/prettier/.stylelintignore、甚至于vite的相关配置也几乎雷同
用 pnpm 还是 leran 去实现这种多项目管理，如何做到好的实现
2.项目整体架构类似与 vben，但比如要把 antd-design 换成其他 ui 框架，如何保留共通的地方，实现快速换 ui 框架

**Comment by Miofly** [comment]
Name: vben 中大量使用了 antd-design 中内置的一些 less 变量，当更换其他 ui 框架时处理这些变量相当麻烦
Role: 有没有好点的方法把使用到 antd-design 中的变量抽离出来，方便替换

**Comment by Miofly** [comment]
Name: 当前的 tab 重新加载页面，会先 router.push 到 redirect/index.vue 页面，然后再通过 replace 到当前刷新的页面，这样会有个 url 变换的过程
Role: 有没有什么特殊的方式能让 url 不变化同时重新加载当前 tab 页面
Online reference example (if available): https://www.qyyjt.cn/cloudTotal
（ps-这个是react开发的项目具有这样的效果）

**Comment by JobinJia** [comment]
关于Form的composable `useForm`
因为本质上vue接受的参数可以是响应式的。那么我们只需要在useForm中把getFormValue换成一个ref对象就可以实现对调用层的双绑。
这样可以不使用setFormValue去单独更新值。
同理。
还可以把`useForm`的参数props设置为MaybeRef<Partial<FormProps>>对象。 这样可以在watchEffect中去判断是参数props是响应式的，则自动调用updateProps去更新Form的props. 在改变props时，可以省掉很多事儿。(可能会有一点点性能上的损失, 因为参数变成了reactiv的)。 如果props是普通对象，则就像现在这样，使用暴露的方法去指定更新就行
那么在使用的时候 可以

```
 const [modelRef, { register }] = useForm<ref({
   schemas: [],
   ...xxxProps
 })>

// 甚至我们可以指定model的类型。比如
type useForm<T> = (props: MaybeRef<Partial<FormProps>>) => {
  modelRef: Ref<T>
  methods: Recordable
}

// 使用时就可以获取 modelRef的类型
```

**Reply by lzdjack** [reply]
现在就是对整个props进行了监听，监听到变化就会setProps呀

**Reply by JobinJia** [reply]
是的，没有错，但实际上如果props整体不是reactivty的并不会生效。这是其一
虽然这样子定义后，`useForm`中的watch可以生效
```
const props = {
  propItem: ref(a)
  propItem2: computed(() => {
    return xxx
 })
}
props.propItem.value = xxx // 这样子可以触发useForm中的watch
```

但是
```
const props = {
  a: 'xx',
  b: 'xx'
}

props.a = xx // 这样并不能
```
其二 就是 formModel的object对象，最好是响应式的，这样可以省去自己去getFormValue。而且在更新值的时候
可以 直接 设置 。 如

```
const { modelRef } = useForm<{
  a: string
  b: number
}>(props)

// 如果需要改变form中的值， 则
modelRef.value.a = 'string'
modelRef.value.b = 2
```
